internal/pkg/propagation/stdlib.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagation
+
+import (
+	"go/types"
+	"strings"
+
+	"github.com/google/go-flow-levee/internal/pkg/utils"
+	"golang.org/x/tools/go/ssa"
+)
+
+// taintStdlibCall propagates taint through a static call to a standard
+// library function, provided that the function's taint propagation behavior
+// is known (i.e. the function has a summary).
+func (prop *Propagation) taintStdlibCall(call *ssa.Call, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	summ, ok := funcSummaries[funcName(call)]
+	if !ok {
+		return
+	}
+
+	prop.taintFromSummary(summ, call, call.Call.Args, maxInstrReached, lastBlockVisited)
+}
+
+// taintStdlibInterfaceCall propagates taint through a call to a function that
+// implements an interface function from the standard library, provided that the
+// function's taint propagation behavior is known (i.e. the function has a summary).
+// This can be a static call or a dynamic call.
+func (prop *Propagation) taintStdlibInterfaceCall(call *ssa.Call, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	summ, ok := interfaceFuncSummaries[funcKey{funcNameWithoutReceiver(call), sigTypeString(call.Call.Signature())}]
+	if !ok {
+		return
+	}
+
+	var args []ssa.Value
+	// For "invoke" calls, Value is the receiver
+	if call.Call.IsInvoke() {
+		args = append(args, call.Call.Value)
+	}
+	args = append(args, call.Call.Args...)
+
+	prop.taintFromSummary(summ, call, args, maxInstrReached, lastBlockVisited)
+}
+
+func (prop *Propagation) taintFromSummary(summ summary, call *ssa.Call, args []ssa.Value, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
```

Comment:
Making sure I understand: you are saying that `taintStdlibInterfaceCall` can cover both static and dynamic calls to interface methods? If so, that is my intent. (Maybe the docstrings are unclear on this point?) Specifically, given a `funcKey` such as `{"Write", "([]byte)(int,error)"}`, I want to match both of the following calls:
```
func foo(s core.Source, r io.Writer, rr myTypeThatImplementsWriter) {
  r.Write([]byte(s.Data))  // dynamic call to Writer.Write
  rr.Write([]byte(s.Data)) // static call to myTypeThatImplementsWriter.Write 
}
```

If instead you were saying that a function call might lead to `taintFromSummary` being called both through `taintStdlibCall` and `taintStdlibInterfaceCall`, unless I messed up, there should not be any overlap between the summaries for these two handlers, so that should not happen.

For the `logger` example you give, I don't think that would match any summaries, so there shouldn't be an issue?

Reply:
My initial comment was mostly a reminder to myself to review that the difference between `Args[0]` being the receiver or not is, in fact, exactly identified by `IsInvoke()`.  I remembered that being a stumbling block in the past, but it does look like you've got it well covered in this implementation.

Marking this chain resolved.